fat_error_decrypt.go

```diff
@@ -0,0 +1,201 @@
+				SenderIdx: 1,
+				Sender:    o.circuit.PaymentPath[0],
+			},
+		}, nil
+	}
+
+	sharedSecrets, err := generateSharedSecrets(
+		o.circuit.PaymentPath,
+		o.circuit.SessionKey,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error generating shared secret: %w", err)
+	}
+
+	var (
+		sender      int
+		msg         []byte
+		dummySecret Hash256
+	)
+	copy(dummySecret[:], bytes.Repeat([]byte{1}, 32))
+
+	// We'll iterate a constant amount of hops to ensure that we don't give
+	// away an timing information pertaining to the position in the route
+	// that the error emanated from.
+	hopPayloads := make([][]byte, 0)
+	for i := 0; i < o.maxHops; i++ {
+		var sharedSecret Hash256
+
+		// If we've already found the sender, then we'll use our dummy
+		// secret to continue decryption attempts to fill out the rest
+		// of the loop. Otherwise, we'll use the next shared secret in
+		// line.
+		if sender != 0 || i > len(sharedSecrets)-1 {
+			sharedSecret = dummySecret
+		} else {
+			sharedSecret = sharedSecrets[i]
+		}
+
+		// With the shared secret, we'll now strip off a layer of
+		// encryption from the encrypted error payload.
+		encryptedData = onionEncrypt(&sharedSecret, encryptedData)
+
+		message, payloads, hmacs := o.getMsgComponents(encryptedData)
+
+		expectedHmac := o.calculateHmac(
+			sharedSecret, i, message, payloads, hmacs,
+		)
+		actualHmac := hmacs[i*sha256.Size : (i+1)*sha256.Size]
+
+		// If the hmac does not match up, exit with a nil message.
+		if !bytes.Equal(actualHmac, expectedHmac) && sender == 0 {
+			sender = i + 1
+			msg = nil
+		}
+
+		// Extract the payload and exit with a nil message if it is invalid.
+		payloadType, payload, err := o.extractPayload(payloads)
+		if sender == 0 {
+			if err != nil {
+				sender = i + 1
+				msg = nil
+			}
+
+			// Store data reported by this node.
+			hopPayloads = append(hopPayloads, payload)
+
+			// If we are at the node that is the source of the error, we can now
+			// save the message in our return variable.
+			if payloadType == payloadFinal {
+				sender = i + 1
+				msg = message
+			}
+		}
+
+		// Shift payloads and hmacs to the left to prepare for the next
+		// iteration.
+		o.shiftPayloadsLeft(payloads)
+		o.shiftHmacsLeft(hmacs)
+	}
+
+	// If the sender index is still zero, all hmacs checked out but none of the
+	// payloads was a final payload. In this case we must be dealing with a max
+	// length route and a final hop that returned an intermediate payload. Blame
+	// the final hop.
+	if sender == 0 {
+		sender = o.maxHops
+		msg = nil
+	}
+
+	return &DecryptedFatError{
+		DecryptedError: DecryptedError{
+			SenderIdx: sender,
+			Sender:    o.circuit.PaymentPath[sender-1],
+			Message:   msg,
+		},
+		Payloads: hopPayloads,
+	}, nil
+}
+
+const (
+	payloadFinal        = 1
+	payloadIntermediate = 0
+)
+
+func (o *OnionFatErrorDecrypter) shiftHmacsLeft(hmacs []byte) {
+	if len(hmacs) != o.allHmacsLen {
+		panic("invalid hmac block length")
```

Comment:
return error instead?

Reply:
Check removed, because length checking is already carried out at a higher level.